whitesource/component.py

```diff
@@ -84,3 +89,136 @@ def download_component(
+            elif filter.type == 'resource':
+                c.resources = artifacts
+            logger.debug(f'past filter ({filter.type=}): {len(artifacts)} artifacts')
+            yield c
+
+        elif isinstance(m := filter.match, dict):
+            tmp = []
+            for a in artifacts:
+                if isinstance(a.type, cm.ResourceType):
+                    resource_type_name = a.type.value
+                elif isinstance(a.type, str):
+                    resource_type_name = str(a.type)
+                else:
+                    raise NotImplementedError
+
+                logger.debug(f'parsed {a.type} to {resource_type_name=}')
+
+                if filter.action == 'include':
+                    if re.search(
+                            pattern=m.get('type') if filter.type == 'resource' else m.get('name'),
+                            string=resource_type_name if filter.type == 'resource' else a.name,
+                    ):
+                        tmp.append(a)
+                elif filter.action == 'exclude':
+                    if not re.search(
+                            pattern=m.get('type') if filter.type == 'resource' else m.get('name'),
+                            string=resource_type_name if filter.type == 'resource' else a.name,
+                    ):
+                        tmp.append(a)
+            if filter.type == 'source':
+                c.sources = tmp
+            elif filter.type == 'resource':
+                c.resources = tmp
+            logger.debug(f'past filter ({filter.type=}): {len(artifacts)} artifacts')
+            yield c
+        else:
+            raise NotImplementedError
+
+
+def _apply_artifact_action(
+    action: str,
+    artifacts,
+    match: bool,
+):
+    if action == 'include':
+        return artifacts if match else ()
+    elif action == 'exclude':
+        return artifacts if not match else ()
+    else:
+        raise RuntimeError
+
+
+def _apply_component_filter(
+    filter: whitesource.model.WhiteSourceFilterCfg,
+    components: typing.Generator[typing.Union[tuple, typing.Any], typing.Any, typing.Any],
+):
+    if isinstance(m := filter.match, bool):
+        yield from _apply_components_action(action=filter.action, components=components, match=m)
+
+    elif isinstance(m := filter.match, dict):
+        for c in components:
+            if filter.action == 'include':
+                if re.search(pattern=m.get('name'), string=c.name):
+                    yield c
+            elif filter.action == 'exclude':
+                if not re.search(pattern=m.get('name'), string=c.name):
+                    yield c
+    else:
+        raise NotImplementedError
+
+
+def _apply_components_action(
+    action: str,
+    components,
+    match: bool,
+):
+    if action == 'include':
+        return components if match else ()
+    elif action == 'exclude':
+        return components if not match else ()
+    else:
+        raise RuntimeError
```

Comment:
I think a `NotImplementedError(action)` would be more fitting/helpful, especially if enums are used.